### ISSUE TITLE
Play Spotify

### ISSUE THREAD

**Issue by infomobi** [MiczFlor/RPi-Jukebox-RFID#18]
Hello,

thanks for the great work.
Will it be possible to play songs or playlists from Spotify premium?

Thanks four your help!

**Comment by tullm** [comment]
Hi.
Yes this is possible. I've done it. 
I've installed mopidy which included mpd. 
I can trigger to play spoitfy content over mpd with the swipe of a RFID tag.
So install mopidy with the Spotify extension. When this is done and you are able to play Spotify over mopidy it is only some editing of the script here to get it working with RFID. I can share what I've edited.
Cheers

**Comment by infomobi** [comment]
Thanks for your answer.
Would it be possible to make a little tutorial with all steps that I have to do (similar to the install guide of RPI-Jukebox-RFID)?
That would be very helpful :-)

**Comment by tullm** [comment]
Hi. Don't remember it step by step, but have a look here: https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/5#issuecomment-358655801 there is a link to the tutorial that I used to install it.
Good luck...

**Comment by infomobi** [comment]
OK, installation is not the problem with the tutorial.
But how can I configure triggering a playlist or album with RFID?

**Comment by tullm** [comment]
Give me a day or two for the answer. Will get back to you here...

**Comment by tullm** [comment]
So,

if you have installed mopidy and spotify premium ist working for you you can go on like that:

I've crated a variable in the "rfid_trigger_play.sh" under /home/pi/RPi-Jukebox-RFID/scripts like that:

`CMDMUSIK1="0001234567` (the number represents the number from the rfid card.

Then i added another command (right under all the other commands)in the same script like that:
`elif [ "$CARDID" == "$CMDMUSIK1" ]`
`then`
`    # stop VLC and Start SpotifyPlaylist Musik1`
`    mpc -p 6681 stop && mpc -p 6681 clear && mpc -p 6681 load Musik1 && mpc -p 6681 volume 80 && mpc -p 6681 play && sudo pkill vlc
` <- this whole mpc stuff is one line

So when you swipe the card 0001234567 it detects the special card which triggers the command. the commant stops mpc if it is playing, clears the current playlist, loads the Playlist Musik1 from spotify, sets volume to 80%,starts playing and kills vlc. You can load all playlists here by name. Or with the us of the uri url from spotify. 

All command available are listed here: https://linux.die.net/man/1/mpc 

I also made it possible to trigger paue, next and so on by adding these command to the normal command within the script like that:

`elif [ "$CARDID" == "$CMDPAUSE" ]`
`then`
`    # pause VLC media and mpc`
`    echo "pause" | nc.openbsd -w 1 localhost 4212`
`    mpc -p 6681 toggle`

enjoy.

hope that helps.
tullm

**Comment by infomobi** [comment]
Thank you, that helped a lot and now it works fine.

I made a little addition to the code to make everything a little more flexible. I can add new spotify playlists to the jukebox without editing the script. I just have to create a new audiofolder with a file named playlist.spotify. The content of this file is the name of the playlist.

Here is the code in rfid_trigger_play.sh:

```
        if [ -f $PATHDATA/../shared/audiofolders/$FOLDERNAME/playlist.spotify ]
        then
                playlistname=`cat $PATHDATA/../shared/audiofolders/$FOLDERNAME/playlist.spotify`
                sudo pkill vlc && mpc clear && sleep .1 && mpc load $playlistname && sleep .1 && mpc play
        else
```
You just have to insert it in the right place:
```
    # if a folder $FOLDERNAME exists, play content
    if [ -d $PATHDATA/../shared/audiofolders/$FOLDERNAME ]
    then
INSERT HERE
                # create an empty string for the playlist
                PLAYLIST=""
```
Again, thanks for your help.

PS: Sometimes, spotify stops working on my machine. The logfile shows that loading a playlist runs into a timeout. Do you have the same problem?

**Comment by MiczFlor** [comment]
I am planning to add the `spotify` option to the script in core. However, since I have no spotifiy (...) can somebody give me a brief "install" info here for the initial steps to get the Pi ready for spotify? Or should I just refer to the link posted by @tullm ?
> To add spotify-connect I used: https://github.com/dtcooper/raspotify
> tutorial to install is on github there.

**Comment by tullm** [comment]
@MiczFlor Hi, great. But the link you postet is for spotify connect, not for spotify itself.

this ist the right one:

> To add the Mopidy Music Server (for local or networkfiles, spotify and other music services) I used: https://www.mopidy.com/
tutorials:
english: https://maex.me/2017/10/raspberry-pi-as-a-music-server/
german: https://braspi.de/blogs/braspi-blog/raspberry-pi-als-musik-server-mit-mopidy
or the very good documentation from mopidy: https://docs.mopidy.com/en/latest/
   I used extensions: spotify, local-images (cover database for local or network files), iris (webinterface), local files.

If you need help or testing let me know. (also in german if neede ;-) )

@infomobi ohh, that sound very interesting. Will try it. The Foldername ist the cardnumber then?

**Comment by infomobi** [comment]
The foldername in audiofolders is an alias like "Karneval" and includes the file playlist.spotify with the name of the playlist as content.
The automatically created folder CARDID in folder shortcuts points to this folder like it does with normal mp3-playlists.

@MiczFlor I can try to write the "step-by-step install" for spotify in the next days. You have to add my extension of your script to rfid_trigger_play.sh then.

**Comment by princemaxwell** [comment]
@MiczFlor Hi, Spotify Connect is not needed. I have integrated spotify into my Pi by using Mopidy + MPC.
To get this working i created two options: Normal mode and Shuffle Mode.
I need one file in my audiofolder: "spotify.txt" or "spotifyshuffle.txt". Into that file i have to put the URL of the spotify playlist.


PS:
I made some improvements to the webinterface too:
-Play button is working for spotify
-Playing song is shown
-Covers integrated to the UI
-Added + and - buttons for Volume
-Added a filemanager to directly edit/delete the shortcuts and upload/edit/delete files to audiofolders
-Added direct links to the UI for editing/deleting shortcuts and audiofolders
-Some improvements for user friendlyness

Other improvements to rfid_trigger_play.sh:
-Added some association options to cards (shuffle, reboot, volume up, volume down ....)
-spotify integration
-shuffle for VLC and spotify
-automatic audiofolder creation (if shortcut points to an audiofolder which doesn't exist, it will be automaticalle created by reading the card)

**Comment by MiczFlor** [comment]
hi @princemaxwell thanks for your input. If you were to share your work here or as a pull request, I would like to review and include it, if that is possible? And please don't forget to post a picture of your jukebox in the other thread to be included in the image gallery ;)

**Comment by princemaxwell** [comment]
@MiczFlor sure, thats possible.
i will send you the files on evening...

**Comment by infomobi** [comment]
Would you publish the files for testing here @MiczFlor and @princemaxwell?
I would love to test everything.

**Comment by princemaxwell** [comment]
I have to remove some very individual changes from my code, which are only affecting my system.
And i have to add some comments...

**Comment by MiczFlor** [comment]
@princemaxwell ok, thanks for the feedback. Looking forward to the code.

**Comment by infomobi** [comment]
@princemaxwell Any progress here? I would love to test everything. Thanks!

**Comment by MiczFlor** [comment]
Hi @nlssttzl 
Good point. This thread needs to go into the documentation for the branch with the old vlc code. 
And if nobody else ventures into the new mpd version, I might need to get a Spotify account to figure   out. It will take a few weeks. Busy with consolidating a few improvemebts and other tickets. 
But thanks for reminding me. It will come. 
All the best, micz

**Comment by nlssttzl** [comment]
Actually this shouldnt be too much of a problem since all the commands still work. So via console it ia still possible to steer Spotify. I should even get easier since soy just have to worry about mpc and not vlc. 
Just have to figure out where and how to integrate it into your script and start playing once the card is being swiped. 
Too bad I’m a really bad at programming. Maybe I’ll spend 1-2 evenings figuring out and trying around. However I would be glad if someone shares their experience.

**Comment by nlssttzl** [comment]
So I`m back from bouncing some ideas back and forth and I managed to get it running. Unfortunatly not with the more easy version of @infomobi but nevertheless, better than nothing. It should be possible to include the whole play/stop as well, but here finally my noob-skills are too noob.
In addition I did not manage yet to stop spotify when playing a "regular" card.
Well here's what I got:

First link a command to the card swiped in the settings/rfid_trigger_play.conf:
`cmdplayspotify="0002763033"`
Whereas the number matches the RFID-tag.

After that create in the scripts/rfid_trigger_play.sh in the section where all the case commands are (after: `case $CARDID in`

`       $cmdplayspotify)
         mpc stop ;
         mpc clear;
         mpc -p 6601 stop;
         mpc -p 6601 clear;
         mpc -p 6601 load auto && mpc -p 6601 play
        ;;
`
In that case "auto" is my spotify list. That command stops all "regular cards" and all running "spotify cards". 6601 is in my case the port which is being defined in the mopidy.conf

That's it. Not much but in a first step it's running. Next step now would be to include start/stop and to so playing spotify once a "regular card" is being swiped.

**Comment by no-banker** [comment]
First thanks a lot for the great work and the great idea! My first steps are done but now I want to integrate Spotify and I read this post. 

@MiczFlor If I understood right you want to integrate the Spotify support in one of the next versions. You have an idea when you start with the implementation? I ask because If you are right now working on it I don’t start with the dirty version if the integration is already in progress. If you need help by testing this feature and if I’m able to help feel free to contact me.

**Comment by MiczFlor** [comment]
Hi @no-banker 
thanks for offering support. A "working hack" is all we need at this stage for the latest phoniebox version. 
How would you approach the issue?
If you get started on this, could you have a look at http://tizonia.org/ ?
My thoughts were to use this software as part of phoniebox. It it works, I would then add it to the install script.

**Comment by no-banker** [comment]
Hi @MiczFlor 

Tonight I had time to test with tizonia. Seems to be a propper solution. But you have to install several tools on top:

`pip install spotipy

sudo apt-get install pulseaudio`

I have tried to do everything without config file in tizonia by using the following command

`tizonia --spotify-user myusername --spotify-password mypwd --spotify-playlist nameofplaylist`

Afterwards you can steer with the keyboard:

Keyboard control:

   [p] skip to previous file.
   [n] skip to next file.
   [SPACE] pause playback.
   [+/-] increase/decrease volume.
   [m] mute.
   [q] quit.

These are all options for spotify which you can use:

Spotify options (Spotify Premium required):
  --spotify-user arg     Spotify user name  (not required if provided via 
                         config file).
  --spotify-password arg Spotify user password  (not required if provided via 
                         config file).
  --spotify-owner arg    The owner of the playlist  (optional: may be used in 
                         conjunction with --spotify-playlist).
  --spotify-tracks arg   Search and play from Spotify by track name.
  --spotify-artist arg   Search and play from Spotify by artist name.
  --spotify-album arg    Search and play from Spotify by album name.
  --spotify-playlist arg Search and play public playlists (owner is assumed 
                         current user, unless --spotify-owner is provided).

Hopefully you can integrate it with this information? Sorry but I don’t have the skills to write the code.

**Comment by bgraeb** [comment]
From what I can see, tizonia would not help at what we want to achieve, as it has no headless mode (right now). 

I did a basic implementation with `mopidy`, but I had no time to integrate it.

This starts a playlist from Spotify:

```python
from mpd import MPDClient

client = MPDClient()
client.timeout = 200
client.idletimeout = None
client.connect("localhost", 6600)
client.add("spotify:user:d5l7k1dqq59jr1qrtdlicykiq:playlist:1zVlcCYZ4BII9D7KMjceJv")
client.play()
```

As you can see, you can run this headless. One could use `sys` to read from stdin, as far as I can remember.

You need a registration at Spotify to create a authentication token: [See here](https://github.com/mopidy/mopidy-spotify).

I need to upgrade my current installation. When I'm done, I'll take a look If I can integrate it (haven't had time to look into this project for a while - dunno how far you got).

**Comment by Alwiz79** [comment]
@bgraeb this seems to be in general a pretty interesting thread you linked there.

https://github.com/mopidy/mopidy-spotify/issues/182#issuecomment-410465484

I also think that Tizonia won’t help us. In the above thread someone mentions he uses NFC cards that hold a Spotify uri to be played and passes that on to mopidy :-)
And your first trials look good.
I’ll return from holidays in around two weeks, will try to help then if it’s not solved until then

**Comment by MiczFlor** [comment]
Hi @bgraeb @no-banker @Alwiz79 
looking forward to get a Spotify account to test your pull requests :)
My hope for tiziona was that it would be a neat wrapper around not only Spotify but Google Music and other services. The issue at hand: is it "truely" headless. I haven't tried it.
Please take a look also at an older thread: https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/118#issuecomment-410885541
There is some information on Spotify - which, as you can imagine without an account (yet!) - I haven't tested.
Keep the good thoughts and contributions coming!
all the best, micz

**Comment by Alwiz79** [comment]
@bgraeb did you install mopidy in _parallel_ to the existing jukebox project or _instead_ for this basic implementation?

I am asking because:
I installed mopidy in _parallel_. When I want to start mopidy I get an error message saying:
`Frontend (MpdFrontend) initialization error: MPD server startup failed: [Errno 98] Address already in use`
That is probably because installing in parallel is a stupid idea? Not sure, too noob for that.

anyhow it also says: "Logged in to Spotify in online mode" so..I guess the first step is done :-)

**Comment by bgraeb** [comment]
The error you see is, because mpd is a server component that we launch and use. So, it starts and tries to register a listener on your system. If there is already a listener on your system on that port, you'll see this error. So, no panic.

Maybe, I'll get some time this week to do an implementation of the Spotify part.

**Comment by Alwiz79** [comment]
@MiczFlor just signed up for spotify premium account. But now I have to admit, I don't know how to move on. If you need access to premium account, let me know (PM?).

**Comment by bgraeb** [comment]
@Alwiz79: Next step would be, getting an [authentication token](
https://github.com/mopidy/mopidy-spotify). With that you can automate the playback of Spotify with mpd.

Then, the playback of a Spotify playlist must be integrated. From what I have seen (buying a house breaks all your plans regarding hobbies and development  :-/ ): 

- they need to be recognized as a playlist
- The app should create a config from it
- the config must be used to playback the playlist

**Comment by Alwiz79** [comment]
got that authentication token, and as mentioned before: when i start mopidy it says:
"Logged in to Spotify in online mode"
..besides the error message.
I guess i simply don't understand (enough) how to control mopidy..does this thing need some kind of client? or command line input? i'm totally confused about that parallel installtion and how to move on

@bgraeb: if you think buying a house brings chaos to your life, wait for children :-)...but after a few years you will have time for your hobbies again (look at me)...anyhow i don't want to wait years for the spotify integration :-)

**Comment by Alwiz79** [comment]
I found the foloowing issue:
[https://github.com/mopidy/mopidy-spotify/issues/140](https://github.com/mopidy/mopidy-spotify/issues/140)
This might be the reason, why I didn't manage to load a playlist via mpc command line

**Comment by princemaxwell** [comment]
I checked Spotify integration for Version 1.1.4 after i succesfully integrated in the old VLC Version < 0.94.
The integration with mopidy was very easy for the old VLC version, but i got many problems for version 1.1.4 because it uses MPD on port 6600. Mopidy uses MPD on port 6600 by default too. I could change the port to 6601, so that two MPD servers are running.

There are only 5 steps to get this running. The syntax for MPC is then "mpc -h 0.0.0.0 -p 6601 add spotify:album:###"

Before that i told the webinterface of phoniebox, that type Spotify has to create a spotify.txt with the spotify URL in it.

My problem at the moment is, to get this spotify url from spotify.txt and start playing by using mpc on port 6601.

In the actual phoniebox code, port 6600 is hardcoded so that there is no easy way to get all the commands working on port 6601.

if someone wants a description, how to get mopidy working including mpc and has the time to get this all together in the code, i will paste my self-written manual here.

**Comment by princemaxwell** [comment]
> 
> 
> I found the following issue:
> [mopidy/mopidy-spotify#140](https://github.com/mopidy/mopidy-spotify/issues/140)
> This might be the reason, why I didn't manage to load a playlist via mpc command line.
> 
> And this seems to be an approach to fix it. Need to take a deeper look at it.
> [mopidy/mopidy-spotify#188](https://github.com/mopidy/mopidy-spotify/pull/188)

yes, it's sure.
i tested it in my fully working VLC version with spotify.
Tracks, Albums are playing, but playlists aren't working anymore. In April 2018 they worked, and i have not changed anything since then...

We have to hope, that there is a workaround for mopidy-spotify...

**Comment by Alwiz79** [comment]
But there seems to be a working Fix for that.
I just don’t know how to implement this (yet).

And this seems to be an approach to fix it. Need to take a deeper look at it.
mopidy/mopidy-spotify#188

**Comment by Alwiz79** [comment]
can anybody who has the coding skills please take a look at the spotify integration of playlists. I'm really stuck here because of my lack of coding skills, but i urgently need spotify to  be running (birthday present of the phoniebox).....

**Comment by kathateich** [comment]
The project https://github.com/fsahli/music-cards supports Spotify as well as local files. I don't know python good enough to say if there are ideas that are helpful for this Jukebox. But maybe someone else knows?!
I am interested in support for Spotify, too.

**Comment by MiczFlor** [comment]
Hi @kathateich 
the code of this project has not been touched in three years, for all I know. I have to check again, but I am sure it will run into the same problems listed in this thread. The problems come from the lack of support by Spotify to their API which would allow to get info and sound from their platform into other platforms.

I feel with you guys (like @Alwiz79 put it in a few posts above) and would like to dedicate time to this. From experience I know that opening this can of worms requires an entire weekend of work. Firstly to make it work (might only take a few hours) to fixing all the triggered glitches and problems it will cause (takes three times the original time put in) to finally documenting it (another hour). And then merging it - and catching the debris of the changes for the coming days, posted in these threads here.

This is how it works and it is a good process. My hope is somebody else makes the first steps and I can focus on documentation and bug fixing :) simply because right now I don't have the time for this hike and focus on smaller walks.

**Comment by princemaxwell** [comment]
> 
> 
> But there seems to be a working Fix for that.
> I just don’t know how to implement this (yet).
> 
> And this seems to be an approach to fix it. Need to take a deeper look at it.
> [mopidy/mopidy-spotify#188](https://github.com/mopidy/mopidy-spotify/pull/188)

i took a deeper look at this fix.
i might fix the playlist problems for web API and web UI but not for playing spotify playlist over MPC.

But MPC is what we need for handling with @MiczFlor Phoniebox and RFID cards.

BTW: mopidy-spotify is not needed for playing song, artists or albums with MPC.
i removed it and made a reboot. after that i tried playing albums with MPC and it just worked. Without having mopidy-spotify installed. So, thats the reason, why the fix doesn't take effect for MPC.


when i type a command like this:

"mpc clear"
"mpc add "spotify:artist:0oLWXBuoO0namCRC2oji0q"
"mpc play"

this works very fine!

the only thing, whats not working:

"mpc add "spotify:playlist:37i9dQZF1DWUVpAXiEPK8P"
error adding spotify:playlist:37i9dQZF1DWUVpAXiEPK8P: directory or file not found

i don't have an idea, how MPC is interpeting the spotify links.
and i habe no idea, what package has to be updated to understand playlists.
until april 2018 it worked fine...

**Comment by Alwiz79** [comment]
@princemaxwell there are definitly users reporting, that a fix introduced by @BlackLight works,
He has an unofficial fork for mopidy-spotify.
Check the installation guide here: 
[https://github.com/mopidy/mopidy-spotify/issues/182#issuecomment-412370556](https://github.com/mopidy/mopidy-spotify/issues/182#issuecomment-412370556)

Anyhow I could not manage to load playlists. Maybe I simply used the syntax wrong, not sure.
As soon as i load a playlist with the iris frontend, i can then play/pause/stop with the mpc command line.

I tried add and load with play list name, https-link, uri format...nothing. but, as mentioned before there are people who were succesfull

**Comment by princemaxwell** [comment]
I installed the fork from @BlackLight.
Mopidy-Spotify package is NOT needed for MPC command line. I removed it and MPC still worked with
MPC
-play
-add (Spotify:artist/album:###)
-clear
-etc...

Have you tried my syntax for artist link above?
Their you can test syntax.

For the moment only albums, tracks and artists  are possible for phoniebox...

**Comment by Alwiz79** [comment]
@princemaxwell I'm not so quite sure.
I had the original mopidy-spotify extension installed. In addition I installed the Iris frontend to have a nice frontend to play with.
My spotify playlists worked with that (no Blacklight fork installed yet).
But command line wasnt working for playlists.

then I installed the Blacklight fork, but it still didn't work as mentioned above. So, I'm not sure that this fork is needed for the web UI (?)

I now tried the following:
mpc -h 192.168.178.58 -P mopidy -p 6681 add spotify:album:5KD3xJ04ta4XyrBaqpOwe2
mpc -h 192.168.178.58 -P mopidy -p 6681 play
It's not working here. Mpc still plays the last playlist that I loaded with the web UI Iris

**Comment by princemaxwell** [comment]
You have to clear first. Otherwise you add the new album at the end of your existing list.

mpc -h 192.168.178.58 -P mopidy -p 6681 clear

Then add and play like above. That looked good...

**Comment by Alwiz79** [comment]
@princemaxwell I can now confirmwhat you said :-(
after clear, add, play I can play albums, artists etc, but I cannot load a playlist.
But I'm still convinced that I am / we are doing something wrong.
See the comment here:
[https://github.com/mopidy/mopidy-spotify/issues/182#issuecomment-392593817](https://github.com/mopidy/mopidy-spotify/issues/182#issuecomment-392593817)
I interpret this as: the fix should also be working for mpc command line..

BTW: is there a difference between user created playlist, and the spotify own playlists. like the ones for one special music genre e.g.?

**Comment by kathateich** [comment]
@alwiz
I cannot get mopidy work, i.e. when I run mpc via ssh I get error messages
that mpd refuses connection. This always appears after I used mpc add for
the first time.
Would you help me, please?
Do you run it as a service or "normal"?
How do your [http] and [mpd]-sections look like? I guess that somewhere
there is my fault.
Thanks a lot.

Am Samstag, 13. Oktober 2018 schrieb Alwiz79 :

**Comment by Alwiz79** [comment]
@kathateich sorry i can’t copy my sections here - i just left for a vacation this morning.

Anyhow, I’m running mopidy as a service. The configuration file when running as a service is on a different location (compare one of my earlier comments above).

**Comment by princemaxwell** [comment]
@Alwiz79 
I got it! You will never believe, what i had to do...
I used an old client_id and client_secret from developer.spotify.com and i had to renew this on mopidy.com/authenticate/ and then playlists are working FINE.

the whole time i wasn't using the web API ;-) since now!

I now have mopidy working well and will now start merging it into Phoniebox.
Have a nice holiday...

**Comment by Alwiz79** [comment]
Thats good News! Looking forward to the new Spotify-phoniebox:-)

**Comment by princemaxwell** [comment]
@Alwiz79 @MiczFlor 
It's done! I have a first working version of Phoniebox with spotify support ;-)
The best is, that i deactivated MPD, because i preffered to use the MPD of Mopidy, so that not two instances are running. Mopidy MPD is listening on port 6600 by default, so the difference between the master branch and my version are very easy to handle.

I added the following features:
-spotify integration
-cover integration for spotify tracks
-AJAX integration for title, album, artist, time (elapsed and overall)
-playlist with title, album, artist, time
-fix for songs with > 60 minutes. the time now shows hours.
-advanced inc.playerStatus.php

I am testing all the functions now.
I wrote a little step-by-step guide how i got it working ;-)

**Comment by Alwiz79** [comment]
That’s great news! Happy to see the result next week :-)

**Comment by farbschein** [comment]
Thank you. That's great :-)

**Comment by MiczFlor** [comment]
Hi @princemaxwell 
six months after you solved this already for the VLC version of Phoniebox, now *you did it again* :) And thanks to @Alwiz79 to open the hood of the car, too on this issue.
I am waiting for the pull requests and then will merge it with the develop branch for testing. The Spotifiers in the Phoniebox universe: stay tuned for testing!

**Comment by princemaxwell** [comment]
@MiczFlor 
I have made a pull request. You get my installation step-by-step guide, because i hope you can handle the installerscript integration.

**Comment by MiczFlor** [comment]
@princemaxwell 
I merged your work with the `develop` branch. So others can test it, too. I also created a new file for your "howto" information:
https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/docs/SPOTIFY-INTEGRATION.md
Please add it in there and create a pull request.
I would like this to be with the Phoniebox community asap in the `develop` branch, also the documentation, because (in my experience) the challenge to make this universal starts only now - and will take some time. So the more input and testing, the better. (Some of the older Phonieboxers might remember the *second swipe* discussions :)

**Comment by petschni** [comment]
@princemaxwell thanks for the PR - great news! What is the best way to "update" my current installation to the "develop" commit?

**Comment by princemaxwell** [comment]
The develop branch is under testing!!
So if you want to test, make a backup of your SD-card and try (!) to use the document: [
https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/docs/SPOTIFY-INTEGRATION.md](url)
and please read it first and the look, what steps you need...

**Comment by farbschein** [comment]
@princemaxwell  I made a first installation of develop branch but couldn't find where to change the Playlist_Folders_Path variable as mentioned in Couldn't find where to set the Playlist_Folders_Path. Could anyone help?
https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/docs/SPOTIFY-INTEGRATION.md#change-playlists_folders_path-to
Could you please help?

**Comment by princemaxwell** [comment]
Look at your settings folder in your phoniebox installation:
/home/pi/RPi-Jukebox-RFID/settings/

There should be a file named "Playlists_Folders_Path"

Later we put that step into installerscript...

**Comment by farbschein** [comment]
Thank you. That worked out. 
But now I'm wondering where to type in the spotify track-url (eg. spotify:track:6ZAQxzVUL20heVoowCy...)
Is the stream-url the right place? And what type should I choose?
Thank you again.

**Comment by princemaxwell** [comment]
Have you updated to the develop branch?
Then you should see it when you AddNewStream.
Stream type: Spotify
Stream Url: spotify:(user:username:)album/track/playlist:###

**Comment by farbschein** [comment]
Thank you. Its working now :-)
Got my first track played via RFID from spotify 👍

**Comment by princemaxwell** [comment]
GREAT! Love to hear that it's working.
where did you install the mopidy_spotify FIX now?
can you share a git clone URL for direct installation from git??

**Comment by farbschein** [comment]
I used this command for installing the mopidy-spotify fix in directory:
/usr/lib/python2.7/dist-packages/mopidy_spotify/

git clone -b fix/incompatible_playlists --single-branch https://github.com/BlackLight/mopidy-spotify.git

**Comment by princemaxwell** [comment]
did you navigate to the destination directory first?
when I only use:
`git clone -b fix/incompatible_playlists --single-branch https://github.com/BlackLight/mopidy-spotify.git`
it will be installed into /home/pi

**Comment by farbschein** [comment]
maybe something went wrong. Just had a look into the /usr/lib/python2.7/dist-packages/mopidy_spotify/ directory and figured out that there are 3 mopidy_spotify directories as child-directories. So maybe there went something wrong with my git comment.
The strange thing is, that spotify works - even if I probably didn't got the version with the fix...
So better don't use this command until there is a better solution..

**Comment by BlackLight** [comment]
> maybe something went wrong. Just had a look into the /usr/lib/python2.7/dist-packages/mopidy_spotify/ directory and figured out that there are 3 mopidy_spotify directories as child-directories. So maybe there went something wrong with my git comment.

You have other versions installed. You can safely remove all of them and then install my fork:

```
[sudo] apt purge mopidy-spotify
[sudo] rm -rf /usr/lib/python2.7/dist-packages/mopidy_spotify*
[sudo] rm -rf /usr/lib/python2.7/dist-packages/Mopidy_Spotify-*
git clone -b fix/incompatible_playlists --single-branch https://github.com/BlackLight/mopidy-spotify.git
cd mopidy-spotify
[sudo] python setup.py install
```

**Comment by princemaxwell** [comment]
it would even works because you have installed original branch by sudo apt-get install Mopidy-spotify.
Spotify is working here but not with playlist spotify URI

**Comment by MiczFlor** [comment]
To those who want to test the [spotify integration](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/docs/SPOTIFY-INTEGRATION.md), here is how to get and checkout the branch `develop`
# How to switch to the `develop` branch, assuming:

* you can delete your `develop` branch, if you have any

~~~
cd /home/pi/RPi-Jukebox-RFID/
git checkout master
git branch -D develop
git checkout --track origin/develop
~~~

This should end with something like this: 
~~~
Branch 'develop' set up to track remote branch 'develop' from 'origin'.
Switched to a new branch 'develop'
~~~

And you can check if you are on the branch by typing:
~~~
git branch
~~~

To make sure the code is really what you are looking for, typing the following:
~~~
cat htdocs/ajax.loadCover.php | grep spotify.com
~~~

Should return (at least, possibly more if the code has been changed since I wrote this):
~~~
$url = "https://open.spotify.com/oembed/?url=".$playerStatus['file']."&format=json";
~~~

**Comment by MiczFlor** [comment]
@BlackLight 
The commands 
~~~apt purge mopidy-spotify
rm -rf /usr/lib/python2.7/dist-packages/mopidy_spotify*
rm -rf /usr/lib/python2.7/dist-packages/Mopidy_Spotify-*
git clone -b fix/incompatible_playlists --single-branch https://github.com/BlackLight/mopidy-spotify.git
cd mopidy-spotify
python setup.py install
~~~
[Where in the *howto* ](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/docs/SPOTIFY-INTEGRATION.md) would you add them? Just tell me the line, I will do the editing.

**Comment by BlackLight** [comment]
@MiczFlor sudo is only required for the commands that remove or install files outside of your home, so yeah your guess is correct. Also note that apt purge/remove might be required to cleanly uninstall mopidy-spotify if you're on Raspbian/Debian, otherwise the next mopidy update-purge will overwrite the directory.

Feel free to integrate those lines in the `Install MOPIDY` section, replacing all the part that goes from _To install one of the listed package_ until the end (actually advising to install mopidy-spotify via apt isn't a good idea, since this fix hasn't been merged into the official package yet and apt would just install the wrong version).

**Comment by kathateich** [comment]
I tried to install and edit "everything". Spotify works but local music does not.

**Comment by princemaxwell** [comment]
@kathateich You are right!
That's a issue i fixed by changing the Playlists_Folder_Path to "/home/pi/RPi-Jukebox-RFID/playlists/".
If you want to manually fix ist, change it in "/home/pi/RPi-Jukebox-RFID/settings/Playlists_Folder_Path" and in /etc/mopidy/mopidy.conf put into [m3u] section:
```
[m3u]
playlists_dir = /home/pi/RPi-Jukebox-RFID/playlists
```

**Comment by dnkroy** [comment]
Hello
im having Troubles with the install at this step
sudo nano ~/.config/mopidy/mopidy.conf
the file does not exists. what i am doing wrong?

**Comment by kathateich** [comment]
The file ist created when you start mopidy the first time. So either you
start mopidy 8nce or you just create it yourself.

Am Samstag, 20. Oktober 2018 schrieb dnkroy :

**Comment by Alwiz79** [comment]
@princemaxwell could you please explain again what's to do because of the local music problem 
@kathateich just mentioned above?

I am currently having the same problem. spotify works great, but when i want to start local music it just stops.

besides that i am not having any issue with the spotify integration so far. great work!
just a question: is there a way to save allready in the uri, that i want the playlist to be shuffled?
something like :shuffle:ON   :-)

**Comment by princemaxwell** [comment]
@Alwiz79 did you update or fresh install?
In the Spotify installer there should be no problem, only the local music is loaded slower. This problem is under investigation.

For Spotify you can turn on shuffle in the web UI...

**Comment by MiczFlor** [comment]
Hi @princemaxwell 
Please make these two installs parallel for the time beimg. We need time to catch all the bugs and I would like the default install to work as it did - and spotify is a bonus. Otherwsie we are annoying all the non spotify people. 
So the install script should habe copies for mod.conf like „mpd-spotify.conf.sample“
Some things must wirk in both installs, like
* new path to playlists
* m3u8 playlists
... and ... ?
So they need to be tested in bothe versions.

**Comment by Alwiz79** [comment]
@princemaxwell i did a fresh install with the one line script for the spotify version (obviously :-) )

**Comment by Alwiz79** [comment]
@princemaxwell I did a fresh install using the one line command skript for the Spotify version (obviously :-) )

**Comment by Alwiz79** [comment]
@princemaxwell @MiczFlor 
general feeedback on the handling of spotifiy links:
It seems as if the format of a e.g. playlist should be something like 
> spotify:user:spotify:playlist:37i9dQZF1DWUVpAXiEPK8P

My problem is (I can solve it, but probably not every user can) that the link I get out of the spotifiy webplayer or the iris frontend looks something like:
> https://open.spotify.com/playlist/6mIRMg8kpuok7eS3o7uUx7

Maybe I'm just doing it wrong (copying the link), the way it entered in phoniebox and linked to an rfid card could be improved

**Comment by MiczFlor** [comment]
@Alwiz79  @kathateich @dnkroy  
In the `develop` branch, I added the option to change the file ending for playlists. See here for details:
https://github.com/MiczFlor/RPi-Jukebox-RFID/commit/0f31aeb0f5104cb626b0bf4e2795d1e8169b82cd
By default `m3u` is set. Spotify / Mopidy requires `m3u8`.
If you have the time, could you please check if this works for you - and if so, I will merge it with the master branch.

**Comment by kathateich** [comment]
@alwiz: Did you install the fork (Blacklight) in the home directory or the
other path mentioned? In the documentation it says you need to copy them to
a certain directory but it is missing that you need to install it.
I think that that was my fault.

Am Montag, 22. Oktober 2018 schrieb Alwiz79 :

**Comment by Alwiz79** [comment]
@miczflor: this might sound stupid, but I’m really not familiar with github. How do I switch to develop branch?

@miczflor: where did you add the option? During the execution of the installation script (input needed) or somewhere in the phoniebox GUI ?

@kathateich: Not sure..I just used the one line command script for the Spotify version provided by miczflor

**Comment by princemaxwell** [comment]
@Alwiz79 with the Spotify one line script, everything should fine when you have done a fresh installation.

I tested it many times... Have you tried to start music over SSH?

Please use: mpc load "playlistname" && mpc play

What's happening?

**Comment by Alwiz79** [comment]
@princemaxwell when i try to start the music over ssh I get the following:

> pi@PlayPi:~ $ mpc load "chartstuff" && mpc play
loading: chartstuff
volume: 30%   repeat: off   random: off   single: off   consume: off

To me this looks pretty "normal", but still no audio output.

The playlist (chartstuff.m3u8) is located here: /home/pi/RPi-Jukebox-RFID/playlists
The mp3 referenced in the playlist is located here: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/chartstuff

**Comment by princemaxwell** [comment]
Should be ok.
If playlist path is wrong, a error message is shown.
by showing "volume: 30% repeat: off random: off single: off consume: off" the status is changed to play.

Did you check the content of M3U8 in a text editor?
Are the folders and files correct there?

**Comment by Alwiz79** [comment]
This is content of the M3U8

> /home/pi/RPi-Jukebox-RFID/shared/audiofolders/chartstuff/mySong.mp3

**Comment by princemaxwell** [comment]
@Alwiz79 

> This is the content of the M3U8
> 
> > /home/pi/RPi-Jukebox-RFID/shared/audiofolders/chartstuff/mySong.mp3

Should be ok, else!
Damn, can you please check, if "alsasink" is the correct output way for your box?
In mopidy.conf, alsasink is set as standard output for mopidy...
Because it look like the playlist is being played.

Otherwise check the follwing commands (the first is to load and play the playlist):
```
mpc load "chartstuff" && mpc play
mpc
mpc current
mpc playlist

```
**Please post your requests here!!**


> just a question: is there a way to save allready in the uri, that i want the playlist to be shuffled?
> something like :shuffle:ON :-)

You only can shuffle by using the WebUI and press the shuffle button. To integrate this into URI would be a heavy work...
 
> general feeedback on the handling of spotifiy links:
> It seems as if the format of a e.g. playlist should be something like
> 
> > spotify:user:spotify:playlist:37i9dQZF1DWUVpAXiEPK8P
> 
> My problem is (I can solve it, but probably not every user can) that the link I get out of the spotifiy webplayer or the iris frontend looks something like:
> 
> > https://open.spotify.com/playlist/6mIRMg8kpuok7eS3o7uUx7
> 
> Maybe I'm just doing it wrong (copying the link), the way it entered in phoniebox and linked to an rfid card could be improved

![screen2](https://user-images.githubusercontent.com/37151645/47383148-49a21480-d704-11e8-9711-28ebf4d2cfd4.png)

You have to right-click on album, playlist or track in Mopidy IRIS and choose Copy URI.
But attention. You have to add "user:username" to this link manually!!!
The better way is, to use Spotify client on Mac or Windows. There you can copy directly with user.

**Comment by Alwiz79** [comment]
@princemaxwell 
in mopidy.conf alsasink is selected, yes. If this _correct_ I don't know? With this raspi I'm not using a soundcard or anything but the raspi itself. So I guess it is correct.

> pi@PlayPi:~ $ mpc load "chartstuff" && mpc play
loading: chartstuff
volume: 30%   repeat: off   random: off   single: off   consume: off
pi@PlayPi:~ $ mpc
volume: 30%   repeat: off   random: off   single: off   consume: off
pi@PlayPi:~ $ mpc current
pi@PlayPi:~ $ mpc playlist
Amy Macdonald - Don't Tell Me That It's Over.mp3

To me this looks like the music is playing, but I cannot here it.

Concerning the URI: using the spotify client does the trick..I was always using the web player, and the links to playlist look like described above

**Comment by Alwiz79** [comment]
did a fresh install, it's working now! spotify and local music playing.
only issue but I don't care about this: no cover art is shown (even not for spotify). 
But this [my phoniebox](https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/5#issuecomment-427650552) doesn't have a display that is not important to me

**Comment by SmightyCrown** [comment]
i did also a fresh install  on rpi3 with the oneline installer . Local music is playing but spotify isnt working. I authorized my account in  iris , but in in settings/services it shows "Mopidy-Spotify extension is not running - you will not be able to play any Spotify tracks".  When i swipe a rfid card with a spotify url, nothing happens and when i want to play it from the playlist it shows "Mopidy: Failed to add some tracks". The mopidy.conf should be rigth in my opinion. Any idea what to do ?

**Comment by princemaxwell** [comment]
@Alwiz79 yeah, great news!!
@SmightyCrown Do you tried to reboot?
Otherwise try these commands:

```
sudo systemctl stop mopidy
mopidy
```
Then mopidy is not loaded as a service. Please post the output here.

**Comment by SmightyCrown** [comment]
@princemaxwell 
Yes i did some reboots,its  always the same behaviour.
This is the ouput of the command: 

> INFO     Starting Mopidy 2.1.0
INFO     Loading config from builtin defaults
INFO     Loading config from command line options
INFO     Creating dir /home/pi/.cache/mopidy
INFO     Creating dir /home/pi/.config/mopidy
INFO     Creating dir /home/pi/.local/share/mopidy
INFO     Loading config from builtin defaults
INFO     Loading config from command line options
INFO     Creating file /home/pi/.config/mopidy/mopidy.conf
INFO     Initialized /home/pi/.config/mopidy/mopidy.conf with default config
INFO     Disabled extension spotify: Dependency The 'pyspotify>=2.0.5' distribution was not found and is required by the application not found
INFO     Enabled extensions: iris, mpd, http, stream, m3u, softwaremixer, file, local-images
INFO     Disabled extensions: spotify, local
WARNING  Found local configuration errors, the extension has been automatically disabled:
WARNING    local/media_dir must be set.
WARNING  Please fix the extension configuration errors or disable the extensions to silence these messages.
INFO     Starting Mopidy mixer: SoftwareMixer
INFO     Starting Mopidy audio
INFO     Starting Mopidy backends: StreamBackend, M3UBackend, FileBackend
INFO     Creating dir /home/pi/.local/share/mopidy/m3u
INFO     Audio output set to "autoaudiosink"
INFO     Starting Mopidy core
INFO     Starting Mopidy frontends: IrisFrontend, MpdFrontend, HttpFrontend
INFO     Starting Iris 3.28.1
INFO     MPD server running at [::ffff:127.0.0.1]:6600
INFO     HTTP server running at [::ffff:127.0.0.1]:6680
INFO     Starting GLib mainloop
INFO     Creating dir /home/pi/.local/share/mopidy/local-images
INFO     New MPD connection from [::ffff:127.0.0.1]:41644
INFO     New MPD connection from [::ffff:127.0.0.1]:41642
INFO     New MPD connection from [::ffff:127.0.0.1]:41650
INFO     New MPD connection from [::ffff:127.0.0.1]:41654
INFO     New MPD connection from [::ffff:127.0.0.1]:41658

**Comment by BlackLight** [comment]
@SmightyCrown

`INFO Disabled extension spotify: Dependency The 'pyspotify>=2.0.5' distribution was not found and is required by the application not found`

**Comment by princemaxwell** [comment]
@SmightyCrown

You need to install pyspotify.
By the way, update the extension mopidy-spotify too.

sudo apt-get install libspotify12 python-cffi python-ply python-pycparser python-spotify
sudo rm -rf /usr/lib/python2.7/dist-packages/mopidy_spotify*
sudo rm -rf /usr/lib/python2.7/dist-packages/Mopidy_Spotify-*
cd
sudo rm -rf mopidy-spotify
git clone -b fix/web_api_playlists --single-branch https://github.com/princemaxwell/mopidy-spotify.git
cd mopidy-spotify
sudo python setup.py install
cd

**Comment by SmightyCrown** [comment]
@BlackLight 
@princemaxwell 
Thank you for the fast answer. After i installed pyspotiy and updated mopidy-spotify, it works without a error. The only thing is, that i cant add spotify albums to a card, only tracks. 
With spotify:track:XXXX it works, with spotify:albumXXX in the spotify.txt file, mopidy brings an error: invalid id.

**Comment by dnkroy** [comment]
Did YouTube Miss the : after Album?
spotify:albumXXX

**Comment by SmightyCrown** [comment]
Sorry, that was a typo and its not a user made album.  When i try   

*spotify:album:6zeXEvQUwV55xFB8hSE9nT*
or 
*spotify:user:1131479705:album:6zeXEvQUwV55xFB8hSE9nT*

mopidy shows "Could not load track invalid id"

But with a playlist
*spotify:user:1131479705:playlist:5HQNUKBdkSHpDA2n77LFW5*
or a single track
*spotify:track:0lq8zosTqXuSTkhk3ASVvl*
it works. 

So its ok for me.

**Comment by BlackLight** [comment]
@SmightyCrown can you try from command line?

```shell
mpc add spotify:album:6zeXEvQUwV55xFB8hSE9nT
```

This exact command worked for me. Maybe there's a typo in the command you copied or the branch you cloned and installed is not up-to-date?

**Comment by princemaxwell** [comment]
@SmightyCrown 
Did you try it via command line and MPC?

This cannot work, if you used something like 
`mpc load spotify:album:6zeXEvQUwV55xFB8hSE9nT`

The reason is, that MPC load is for files. You had to use "mpc add spotify:track:0lq8zosTqXuSTkhk3ASVvl"  instead.

In Phoniebox we route the playlist, album or track URI into a .m3u playlist file, which can be loaded via "mpc load ..."

**Comment by princemaxwell** [comment]
@Alwiz79 
**### 1.1.8-beta is live on the DEVELOP branch!**

**For everone who wants to test the newest version with some stability and speed optimizations, feel free to do it and give us a feedback please!**

I updated the UPGRADE instructions today, but have to wait that @MiczFlor will merge this into develop.
Until he has done, here is the instruction.

```
cd /home/pi/RPi-Jukebox-RFID
git checkout develop
git fetch origin
git reset --hard origin/develop
git pull
sudo systemctl stop mpd
sudo systemctl stop mopidy

USERNAME=$(sudo grep 'username' /etc/mopidy/mopidy.conf|sed 's/username = //g'|sed 's/"//g'|tr -d "\n")
PASSWORD=$(sudo grep 'password' /etc/mopidy/mopidy.conf|sed 's/password = //g'|sed 's/"//g'|tr -d "\n")
CLIENT_ID=$(sudo grep 'client_id' /etc/mopidy/mopidy.conf|sed 's/client_id = //g'|sed 's/"//g'|tr -d "\n")
CLIENT_SECRET=$(sudo grep 'client_secret' /etc/mopidy/mopidy.conf|sed 's/client_secret = //g'|sed 's/"//g'|tr -d "\n")

sudo cp /home/pi/RPi-Jukebox-RFID/misc/sampleconfigs/mpd.conf.sample /etc/mpd.conf
sudo cp /home/pi/RPi-Jukebox-RFID/misc/sampleconfigs/mopidy-etc.sample /etc/mopidy/mopidy.conf
sudo cp /home/pi/RPi-Jukebox-RFID/misc/sampleconfigs/mopidy.sample /home/pi/.config/mopidy/mopidy.conf

sudo sed -i 's/%spotify_username%/'"$USERNAME"'/' /etc/mopidy/mopidy.conf
sudo sed -i 's/%spotify_password%/'"$PASSWORD"'/' /etc/mopidy/mopidy.conf
sudo sed -i 's/%spotify_client_id%/'"$CLIENT_ID"'/' /etc/mopidy/mopidy.conf
sudo sed -i 's/%spotify_client_secret%/'"$CLIENT_SECRET"'/' /etc/mopidy/mopidy.conf
sudo sed -i 's/%spotify_username%/'"$USERNAME"'/' ~/.config/mopidy/mopidy.conf
sudo sed -i 's/%spotify_password%/'"$PASSWORD"'/' ~/.config/mopidy/mopidy.conf
sudo sed -i 's/%spotify_client_id%/'"$CLIENT_ID"'/' ~/.config/mopidy/mopidy.conf
sudo sed -i 's/%spotify_client_secret%/'"$CLIENT_SECRET"'/' ~/.config/mopidy/mopidy.conf

echo "classic" > /home/pi/RPi-Jukebox-RFID/settings/edition
EDITION=$(grep 'SPOTinstall' /home/pi/PhonieboxInstall.conf|sed 's/SPOTinstall="//g'|sed 's/"//g'); if [ $EDITION == "YES" ]; then echo "plusSpotify"; else echo "classic"; fi > /home/pi/RPi-Jukebox-RFID/settings/edition

sudo apt-get install libspotify12 python-cffi python-ply python-pycparser python-spotify
sudo rm -rf /usr/lib/python2.7/dist-packages/mopidy_spotify*
sudo rm -rf /usr/lib/python2.7/dist-packages/Mopidy_Spotify-*
cd
sudo rm -rf mopidy-spotify
git clone -b fix/web_api_playlists --single-branch https://github.com/princemaxwell/mopidy-spotify.git
cd mopidy-spotify
sudo python setup.py install

sudo reboot
```

**Comment by Alwiz79** [comment]
@princemaxwell I followed your step by step above on my second Rpi hardware. But it didn't work as expected. So, I had a running spotify installation before (i had issues with local files here).

After the steps above, spotify integration gone.
The GUI shows "classic" in the upper left corner, which might be a hint, that it's not the spotify version ?

**Comment by princemaxwell** [comment]
### EVERYONE WITH PROBLEMS STARTING LOCAL PLAYLISTS HAS TO RUN THIS COMMAND ONCE:

Depending of the location of audiofolder (this is for default location):
`sudo chmod -R 777 /home/pi/RPi-Jukebox-RFID/shared/audiofolders/`

### If you have +Spotify Edition (>1.1.8-beta) you have to SCAN MUSIC LIBRARY
-> Renew your database under Folders & Files.

And please upload this file to fix the upload in the WebUI or wait until @MiczFlor has merged it!
https://github.com/princemaxwell/RPi-Jukebox-RFID/commit/500922f1f768d21e334844ab0f2a047a8cddba90

**Comment by shinebar1001** [comment]
As mopidy is spamming the logfiles with logs like these

> Nov  3 23:12:25 phoniebox mopidy[525]: INFO     New MPD connection from [::ffff:127.0.0.1]:6600


I suggest to create a filter in the syslog to protect the SD-Card from wearing out:

`sudo echo ':msg, contains, "New MPD connection from" ~' >/etc/rsyslog.d/mopidy.conf`
`service rsyslog restart`

and change the logging options of mopidy as well:

`sudo sed -i '/INFO/WARNING/' /etc/mopidy/logging.conf`

**Comment by princemaxwell** [comment]
@shinebar1001 
Great! Thank you. I will integrate the fix into the install script and will pull request this...

**Comment by SmightyCrown** [comment]
Is it possible to update from 1.8beta to 1.18rc2 by git pull or would be a clean install the bettet way? With the beta, i have sometimes problems with local files. Also after the workaround with chmod and renewing the database.

**Comment by ndd** [comment]
First of all: thanks for the great work!

I've just installed Phoniebox 1.1.8-rc2 Spotify Edition and everything works fine, except for loading/playing playlists from Spotify. Tracks and Albums are fine:
mpc add spotify:album:4O83oaztVuBlPDrp50Q3mh (works)
mpc add spotify:track:0aHq4pmQth8Ggj5LyFNBn9 (works)

mpc add spotify:user:Spotify:playlist:37i9dQZF1DXcF6B6QPhFDv (doesn't work)
mpc add spotify:user:tomorrowlandofficial:playlist:0yS25E7g9xQZ1Dst5SqUZn (doesn't work)

What am I doing wrong? :)
Or is this Mopidy issue the reason: https://github.com/mopidy/mopidy-spotify/issues/182

**Comment by princemaxwell** [comment]
@SmightyCrown
You can git pull the develop or master branch.
Follow this update guide to get it to the actual version:
https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/18#issuecomment-433195972

and then

https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/18#issuecomment-434883345

Or do a fresh install by install script and chmod the folder.

@ndd
Have you installed with one line script or have you updated? There is a fix for the playlist issues which is included in our version.

**Comment by princemaxwell** [comment]
@ndd
Sorry. I didn't see yesterday. For playlists you have to use

mpc load "playlistname"

Please try this first...

**Comment by ndd** [comment]
@princemaxwell 
With mpc load "playlistname" I can only load local m3u-playlists, which are in the playlists folder...
This would be fine as long as I could reference a Spotify playlist within such a m3u playlist. But as I mentioned before, this does only work for tracks and albums.

Example 1: album.m3u
spotify:album:4O83oaztVuBlPDrp50Q3mh (works)

Example 1: myplaylist.m3u
spotify:user:tomorrowlandofficial:playlist:0yS25E7g9xQZ1Dst5SqUZn (doesn't work)

**Comment by princemaxwell** [comment]
@ndd @shitzovran 

I tested it here. i created a file "test.m3u" in my playlists directory under "/home/pi/RPi-Jukebox-RFID/playlists/".
The file has the follwing content, like you provided:
`spotify:user:tomorrowlandofficial:playlist:0yS25E7g9xQZ1Dst5SqUZn`

```
pi@phoniebox:~ $ mpc clear
volume: 10%   repeat: off   random: off   single: off   consume: off
pi@phoniebox:~ $ **mpc load "test"**
loading: test
pi@phoniebox:~ $ mpc play
volume: 10%   repeat: off   random: off   single: off   consume: off
pi@phoniebox:~ $ mpc status
DJ Rolando - Knights Of The Jaguar
[playing] #1/467   0:04/2:30 (2%)
volume: 10%   repeat: off   random: off   single: off   consume: off
```
So, for me it worked fine.

Please note: It is not possible to user command line to load playlists directly with spotify uri. if you try something like "mpc load spotify:user:blabla:playlist:blabla" this does not work.
You have to put this URI into a m3u file and then you can try
`mpc load "filename"`
Please put filename into quotation marks.

If this doesn't work, you have to put this on command line.

```
sudo apt-get install libspotify12 python-cffi python-ply python-pycparser python-spotify
sudo rm -rf /usr/lib/python2.7/dist-packages/mopidy_spotify*
sudo rm -rf /usr/lib/python2.7/dist-packages/Mopidy_Spotify-*
cd
sudo rm -rf mopidy-spotify
git clone -b fix/web_api_playlists --single-branch https://github.com/princemaxwell/mopidy-spotify.git
cd mopidy-spotify
sudo python setup.py install
```

**Comment by shitzovran** [comment]
I don't know wht's wrong...

i create new Playlists in "/home/pi/RPi-Jukebox-RFID/playlists/"

the first: "Leo Lausemaus - Folge 01.m3u8". The file has the follwing content:
`spotify:album:5N73vwGXol4maS9U6HLp0o`

The Album works fine.
```
pi@raspberrypi:/ $ mpc clear
volume: 75%   repeat: off   random: off   single: off   consume: off
pi@raspberrypi:/ $ mpc load "Leo Lausemaus - Folge 01"
loading: Leo Lausemaus - Folge 01
pi@raspberrypi:/ $ mpc play
volume: 75%   repeat: off   random: off   single: off   consume: off
pi@raspberrypi:/ $ mpc status
Leo Lausemaus - Leo Lausemaus Titelsong
[playing] #1/23   0:15/1:19 (18%)
volume: 75%   repeat: off   random: off   single: off   consume: off
```


the other Playlist "test.m3u8" with  follwing content:
`spotify:user:tomorrowlandofficial:playlist:0yS25E7g9xQZ1Dst5SqUZn`

```
pi@raspberrypi:/ $ mpc clear
volume: 75%   repeat: off   random: off   single: off   consume: off
pi@raspberrypi:/ $ mpc load "test"
loading: test
mpd error: No such playlist
pi@raspberrypi:/ $ mpc clear
volume: 75%   repeat: off   random: off   single: off   consume: off
pi@raspberrypi:/ $ mpc load "test"
loading: test
pi@raspberrypi:/ $ mpc play
volume: 75%   repeat: off   random: off   single: off   consume: off
pi@raspberrypi:/ $ mpc status
volume: 75%   repeat: off   random: off   single: off   consume: off
```

the playlist doesn't want to work. and no change after reinstallation.
Could it be, there is a problem in the 'One line install command' ?

**Comment by shitzovran** [comment]
No, the same Problem with m3u. 
And without using m3u8, some of my local files won't play . 
I just realized, i have the same problem with podcasts from Spotify...

**Comment by princemaxwell** [comment]
@shitzovran 
What version do you use?
Did you upgrade and did you followed ALL STEPS from the upgrade guide to get you version to 1.1.8-rc2??

**Comment by MiczFlor** [comment]
The code has moved on. I am closing this, assuming the later code solved it. :) All the best, micz

**Comment by ndd** [comment]
I've just installed a clean and fresh Raspbian as well as the latest Phoniebox +Spotify Edition (1.1.8) via one line script.
Still having the problems with playlists mentioned above...

Songs and Albums are working fine, no matter if I call them directly or via a m3u-playlist in the playlists folder.
As soon as I try to load a playlist via a m3u-playlist it doesn't work...

@princemaxwell 
I've run all your commands as well...

**Comment by ndd** [comment]
Seems like this is still an "official" mopidy problem:
https://github.com/mopidy/mopidy-spotify/issues/182

**Comment by kathateich** [comment]
I have some problems, too. My playlists sometimes work, but sometimes (I
cannot reproduce this behaviour) I need to push the play button to start
the music. Is that a "solution" for your problem, too?


Am Sonntag, 16. Dezember 2018 schrieb Niels Döring <notifications@github.com

**Comment by Stev-En** [comment]
Hallo,
es sind ja soviel Kommentare und Tickets offen, da versuche ich hier mal mein Glück.
Der Fehler wird öfters mal beschrieben aber, aber nie richtig drauf eingegangen.

Ich habe wie [hier](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/INSTALL-stretch#one-line-install-command) beschrieben die "Phoniebox + Spotify Edition:" installiert, habe es auch schon mehrmals versucht aber immer der gleiche Fehler. Ich kann in der Web-App Musik (.Mp3) vom Rechner einbinden Youtube geht auch ABER **Spotify lässt sich einbinden aber nicht Abspielen**.

Hier wird der Spotify Accout ja gezeigt [https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/master/docs/SPOTIFY-INTEGRATION.md](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/master/docs/SPOTIFY-INTEGRATION.md)
da hab ich eine Frage wie es Richtig ist mit %....% oder ohne?

username = %spotify_username%
password = %spotify_password%
client_id = %spotify_client_id%
client_secret = %spotify_client_secret%

Weil in der Installation schreibt er es ohne %...%, aber beim Passwort so %spotify_password%!
Wird das Passwort nur nicht richtig anzeigt oder müsste es im Klar Text stehen?


Ich wäre auf Hilfe sehr dankbar. Versuch jetzt schon seit Wochen den Fehler zu finden.

-------------------------------------------------------------------------------------------------

Hello,
There are so many comments and tickets are open, I'm trying my luck here.
The error is often described but never really got into it.

I have like [here] (https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/INSTALL-stretch#one-line-install-command) described the "Phoniebox + Spotify Edition:" installed it, have it several times but always tried the same error. I can integrate music in the web app (.Mp3) from the computer Youtube is synonymous BUT ** Spotify can be integrated but not play **.

This is where the Spotify Accout is shown [https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/master/docs/SPOTIFY-INTEGRATION.md](https://github.com/MiczFlor/RPi-Jukebox -RFID / blob / master / docs / SPOTIFY-INTEGRATION.md)
I have a question how it is right with% ....% or without?

username =% spotify_username%
password =% spotify_password%
client_id =% spotify_client_id%
client_secret =% spotify_client_secret%

Because in the installation he writes it without% ...%, but with the password so% spotify_password%!
Is the password just not displaying correctly or should it be in clear text?


I would be very grateful for help. Try to find the mistake for weeks now.

**Comment by bgraeb** [comment]
Die ` %` sind jeweils nur Platzhalter und müssen mit dem PW ersetzt werden. Also aus `username=%spotify_username%`  wird `username=hanSolo@pirateWeb.star`

**Comment by bgraeb** [comment]
I've setup the box just recently by creating an spotify-token with the website somewhere linked here and using the autoinstaller script. It worked just fine.

**Comment by princemaxwell** [comment]
So ist es, das sind Platzhalter und es müssen im Installer die 4 Daten exakt eingetragen werden. Die Platzhalter werden dann von den Daten ersetzt. Bei den meisten funktioniert es ja, ich vermute also irgendwo einen Fehler bei der Eingabe...

**Comment by Silverhawk0705** [comment]
Hallo, 
meine Spotify Integration funktioniert grundsätzlich, jedoch kann ich keine Öffentliche Playlists von Spotify abspielen. 
Aktuell z.B. die Christmas Hits von Spotify mit folgendem Link:
`spotify:playlist:37i9dQZF1DX0Yxoavh5qJV`
Ist dies ein allg. Problem oder läuft da bei mir etwas falsch?

**Comment by bubble1999** [comment]
I am very new with Github and actually do not know where I could help to edit the documentation.
However, I will share my insigths here and hope someone can put them into the code:
1) SPI is not activated automatically, so e.g. MFRC522 does not work. It has to be enabled via 
`sudo rasp-config` and rebooting. Perhaps it would be possible to show up an info-message during the install scrips to be sure to have SPI activated
2) I got confused and mixed up "mopidy client ID & secret" with "spotify developer client ID & secret". Here it would be great to state in the install script, that the mopidy client ID is needed (and perhaps a hint to the web adress [https://mopidy.com/ext/spotify/#authentication](url)
3) I had a lot of trouble getting the spotify streams running. Same issue as decribed here [https://github.com/natumbri/mopidy-youtube/issues/58](url) or here [https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1067](url)
I got it running by going through this procedure: [https://forum-raspberrypi.de/forum/thread/48179-phoniebox-spotify-ich-bekomme-es-nicht-zum-laufen/](url)
I think the main difference from what I did is that I installed the raspian packages for the desktop environment (because I thought that I will not need them on the headles raspberry) However, it seems that there are some drivers hidden which are needed to get Spotify streams running.
I hope this can help to improve this great project.

**Comment by s-martin** [comment]
Hi @bubble1999, please open a new issue, if you have problems.

Questions in old and closed issues are easily overlooked.
